flex_model/core/hook_function.py

```diff
@@ -51,29 +69,89 @@ def default_editing_function(
+        custom editing functions.
+
+    :note: If no editing function is provided for a :code:`HookFunction`, then
+        this is the default editing function.
+
+    :param nn.Module current_module: Submodule instance hooked into.
+    :param Tensor inputs: Activation tensor produced during the forward pass of
+        the :code:`current_module`.
+    :param Namespace save_ctx: Save context pointer where cached data can be
+        accessed or stored.
+    :param nn.ModuleDict: Pointer to trainable modules globally exposed to all
+        :class:`HookFunction` instances.
+
+    :returns: Edited (or not) activation tensor
+    :rtype: Tensor
+    """
     logger.debug(f"Running default editing function on tensor: {inputs.shape}")
     return inputs
 
 
 class HookFunction:
-    """Respondible for frontend hook function via templating or codegen.
-
-    Args:
-        module_name (str):
-            The module name to be hooked into.
-        expected_shape (Tuple[int, ...]):
-            Expected shape of the activation tensor.
-        editing_function (Optional[Callable]):
-            Optional editing function which will be applied to the activation
-            tensor.
-
-            Has signature (current_module, inputs, save_ctx, modules).
-                current_module (nn.Module): Module hooked into.
-                inputs (Tensor): Full activation tensor.
-                save_ctx (Namespace): Global namespace to save state visible
-                    to all hook (and editing) functions.
-                modules (nn.ModuleDict): Global trainable pytorch layers
-                    visible to all hook (and editing) functions.
+    """Function which retrieves/edits activations in a Pytorch `nn.Module`.
+
+    The user provides the :code:`module_name` of the target submodule. The user
+    can optionally pass in an :code:`editing_function` containing arbitrarily complex
+    python code, which will be used to edit the full submodule activation
+    tensor. If certain dimensions of the activation tensor are expected to be
+    sharded over distributed workers, the user must also provide an
+    :code:`expected_shape` hint so the activation tensor can be assembled.
+
+    :var str module_name: Name of the :code:`nn.Module` submodule to hook into.
+    :var expected_shape: Shape of the full activation tensor. Only the
+        dimensions which are sharded need to be provided. Other dimensions
+        can be annotated as :code:`None` and will be auto-completed.
+    :type expected_shape: Tuple[Optional[int], ...]
+    :var editing_function: Function which is run on the full activation tensor
+        and returns some edited function. Global contexts like the
+        save context and trainable modules are available for use in the
+        editing function runtime.
+    :type editing_function: Optional[Callable]
+    :var save_ctx: Global save context that is exposed to the
```

Comment:
We may want to use param type here instead of var so it renders blocks as params instead of variables but honestly not a big deal.

Reply:
nvm they are vars not the init params